NFC: [CodeCompletion] Use enums in bit fields as-is

rintaro opened the pull request:
We don't need to convert them from/to integer

rintaro committed:
[CodeCompletion] Use enums in bit fields as-is

We don't need to convert them from/to integer

rintaro commented:
@swift-ci Please smoke test

ahoppen commented on include/swift/IDE/CodeCompletion.h:
```diff
@@ -668,13 +668,13 @@ class CodeCompletionResult {
+  CodeCompletionOperatorKind KnownOperatorKind : 6;
+  SemanticContextKind SemanticContext : 3;
+  unsigned char Flair : 8;
+  NotRecommendedReason NotRecommended : 4;
```
Do we have `static_assert`s that e.g. `NotRecommendedReason` doesn’t contain more then 2^4 elements?

rintaro replied:
Good idea. I added a commit.

ahoppen reviewed (commented):
Wow, I can’t believe that we didn’t do this before. LGTM, I think it might be worth also `static_assert`ing that the values actually fit inside the specified number of bits.

rintaro committed:
[CodeCompletion] Add static_assert to limit the max values of enums

rintaro commented:
@swift-ci Please smoke test

bnbarham reviewed (approved):
Much nicer! Note that the build is failing on:
```
SwiftEditor.cpp(611,1): error: static_assert failed due to requirement 'sizeof((anonymous namespace)::SwiftSemanticToken) == 8' "Too big"
static_assert(sizeof(SwiftSemanticToken) == 8, "Too big");
```

ahoppen commented on include/swift/IDE/CodeCompletion.h:
```diff
@@ -665,16 +675,18 @@ class CodeCompletionResult {
+  SemanticContextKind SemanticContext : 3;
+  unsigned char Flair : 8;
+  NotRecommendedReason NotRecommended : 4;
+  bool IsSystem : 1;
```
Not really related to this PR but in libSyntax’s `RawSyntax`, I saw measurable performance improvements when not compressing bitfields into the minimum number of bits but having them byte-aligned. I think this was because the CPU doesn’t need to do bitshifts/masks to retrieve the value.

The bits inside `RawSyntax` were probably accessed a lot more frequently than these, but it might be worth investigating if byte-aligning these might also be beneficial performance-wise, even though it increases the memory footprint.

rintaro committed:
Guard a static assert for MSVC

MSVC doens't pack diffrent underlying int types into a bitfield. e.g.

  struct S {
    int a: 1;
    char b: 1;
    int c: 1;
  };

These fields are considered three sparate bitfields.

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test Windows

ahoppen committed:
[SourceKit] Guard static assert by _MSC_VER

ahoppen commented:
@swift-ci Please smoke test Windows

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
I got this to pass by changing
```diff
- #if !SWIFT_COMPILER_IS_MSVC
+ #if !defined(_MSC_VER)
```

AFAICT `SWIFT_COMPILER_IS_MSVC` is only set if the sources are compiled by the visual studio compiler. https://github.com/apple/swift/blob/b42402bd6f1811a2ab43f58082b2c635a599b4da/include/swift/Basic/Compiler.h#L16-L20
But we are building the Swift compiler using clang (at least the build log in [Windows#19400](https://ci-external.swift.org/job/swift-PR-windows/19400/console) is invoking `T:\llvm\bin\clang-cl.exe`). We are checking for `defined(_MSC_VER)` in quite a lot of places of the code base already, so I assume the check can’t be completely wrong.

rintaro commented:
Makes sense! Merged.
